### PR TITLE
GroupBy array-based result rows.

### LINKED ISSUE #8118: GroupBy array based result rows
### Motivation

GroupBy queries internally represent result rows as `MapBasedRow` objects, which have the following two fields:

```java
private final DateTime timestamp;
private final Map<String, Object> event;
```

As a result, we need to do relatively expensive Map put and get operations (typically these are HashMaps or LinkedHashMaps) at many points: when rows are first generated after each segment scan, when they are merged on historicals, when they are serialized and deserialized, and then when they are merged again on the broker.

The overhead is especially noticeable when the resultset of the groupBy query is large.

See also #6389.

### Proposed changes

1. Create a `ResultRow` class that simply wraps an `Object[]` and allows position-based access.
2. Modify the GroupBy query implementation to use ResultRow throughout, rather than Row / MapBasedRow.
3. Add `ObjectMapper decorateObjectMapper(ObjectMapper, QueryType)` to QueryToolChest, to aid in implementing the compatibility plan described in "Operational impact" below. QueryResource would use it so it could serialize results into either arrays or maps depending on the value of `resultAsArray`. DirectDruidClient would use it so it could deserialize results into ResultRow regardless of whether they originated as ResultRows or MapBasedRows. (By the way, the serialized form of a ResultRow would be a simple JSON array.)

### Rationale

Some other potential approaches that I considered, and did not go with, include:

- Creating an ArrayBasedRow that implements `org.apache.druid.data.input.Row` (just like MapBasedRow does). The reason for avoiding this is that the interface is all about retrieving fields by name -- `getRaw(String dimension)`, etc -- and I wanted to do positional access instead.
- Using `Object[]` instead of a wrapper `ResultRow` around the `Object[]`. It would have saved a little memory, but I thought the benefits of type-safety (it's clear what ResultRow means when it appears in method signatures) and a nicer API would be worth it.

### Operational impact

The format of data in the query cache would not change.

The wire format of groupBy results would change (this is part of the point of the change) but I plan to do this with no compatibility impact, by adding a new query context flag `resultAsArray` that defaults to false. If false, Druid would use the array-based result rows for in-memory operations, but then convert them to MapBasedRows for serialization purposes, keeping the wire format compatible. If true, Druid would use array-based result rows for serialization too.

I'd have brokers always set `resultAsArray` true on queries they send down to historicals. Since we tell cluster operators to update historicals first, that means that by the time the broker is updated, we can assume the historicals will know how to interpret the option. Users would also be able to set `resultAsArray` if they want once brokers are updated, and receive array-based results themselves.

So, due to the above design, there should be no operational impact.

### Test plan

Existing unit tests will cover a lot of this. In addition, I plan to test on live clusters, especially the compatibility stuff.

### PR DESCRIPTION
**Pull request by gianm**
Fixes #8118; see that proposal for additional details about the main change.

Other than the GroupBy changes, the main other "interesting" new/changed classes are:

- ResultRow: The array-based result type.
- BaseQuery: T is no longer required to be Comparable.
- QueryToolChest: Adds "decorateObjectMapper" to enable query-aware serialization
  and deserialization of result rows (necessary due to their positional nature).
- QueryResource: Uses the new decoration functionality.
- DirectDruidClient: Also uses the new decoration functionality.
- QueryMaker (in Druid SQL): Modifications to read ResultRows.

These classes weren't changed, but got some new javadocs:

- BySegmentQueryRunner
- FinalizeResultsQueryRunner
- Query

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByQueryQueryToolChest.java [hunk 1]
```diff
@@ -465,9 +513,9 @@ public String apply(DimensionSpec input)
   }
 
   @Override
-  public CacheStrategy<Row, Object, GroupByQuery> getCacheStrategy(final GroupByQuery query)
+  public CacheStrategy<ResultRow, Object, GroupByQuery> getCacheStrategy(final GroupByQuery query)
   {
-    return new CacheStrategy<Row, Object, GroupByQuery>()
+    return new CacheStrategy<ResultRow, Object, GroupByQuery>()
     {
       private static final byte CACHE_STRATEGY_VERSION = 0x1;
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/groupby/GroupByQueryQueryToolChest.java]: Should this version be bumped so off-machine caches don't cause an issue or will the fancy decorated mapper be used in like `CachingQueryRunner` for reading the values from the cache?

File: server/src/main/java/org/apache/druid/server/QueryResource.java [hunk 2]
```diff
@@ -163,7 +164,9 @@ public Response cancelQuery(@PathParam("id") String queryId, @Context final Http
   public Response doPost(
       final InputStream in,
       @QueryParam("pretty") final String pretty,
-      @Context final HttpServletRequest req // used to get request content-type,Accept header, remote address and auth-related headers
+
+      // used to get request content-type,Accept header, remote address and auth-related headers
```
Comment (clintropolis) [inline server/src/main/java/org/apache/druid/server/QueryResource.java]: nit: `.. request content-type,Accept header, ..` -> `.. request content-type, Accept header, ..` 

(i know it was already like that, so imo only fix if you have something else to change).

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByQueryQueryToolChest.java (hunk 1 above)
Comment (gianm) [reply processing/src/main/java/org/apache/druid/query/groupby/GroupByQueryQueryToolChest.java]: Check out the `prepareForCache` method -- the cache format is the same. It was always array-based, and it remains so.

File: server/src/main/java/org/apache/druid/server/QueryResource.java (hunk 2 above)
Comment (gianm) [reply server/src/main/java/org/apache/druid/server/QueryResource.java]: I'll update this if there ends up being something else to change, for now I'll leave it as it was before so CI doesn't get interrupted.

File: processing/src/main/java/org/apache/druid/query/groupby/GroupByQueryQueryToolChest.java (hunk 1 above)
Comment (clintropolis) [reply processing/src/main/java/org/apache/druid/query/groupby/GroupByQueryQueryToolChest.java]: Oops, looking closer I see this now, thanks :+1:

**Pull request merged**